apollo-runtime/src/main/java/com/apollographql/apollo/internal/RealApolloCall.java

```diff
@@ -119,102 +119,93 @@ private RealApolloCall(Builder<T> builder) {
       response = interceptorChain.proceed(fetchOptions).parsedResponse.or(Response.<T>builder(operation).build());
-    } catch (Exception e) {
-      if (canceled) {
-        throw new ApolloCanceledException("Canceled", e);
-      } else {
-        throw e;
-      }
     } finally {
-      tracker.unregisterCall(this);
-    }
-
-    if (canceled) {
-      throw new ApolloCanceledException("Canceled");
+      transitionToTerminated();
     }
-
     if (queryReFetcher.isPresent()) {
       queryReFetcher.get().refetch();
     }
-
     return response;
   }
 
   @Override public void enqueue(@Nullable final Callback<T> responseCallback) {
-    if (!executed.compareAndSet(false, true)) {
-      throw new IllegalStateException("Already Executed");
-    }
-    if (canceled) {
+    try {
+      transitionToActive(Optional.fromNullable(responseCallback));
+    } catch (ApolloCanceledException e) {
       if (responseCallback != null) {
-        responseCallback.onCanceledError(new ApolloCanceledException("Canceled"));
+        responseCallback.onCanceledError(e);
       }
     }
-    tracker.registerCall(this);
-    interceptorChain.proceedAsync(dispatcher, fetchOptions, interceptorCallbackProxy(responseCallback));
+    interceptorChain.proceedAsync(dispatcher, fetchOptions, interceptorCallbackProxy());
   }
 
   @Nonnull @Override public RealApolloQueryWatcher<T> watcher() {
-    return new RealApolloQueryWatcher<>(clone(), apolloStore, tracker);
+    return new RealApolloQueryWatcher<>(clone(), apolloStore, logger, tracker);
   }
 
   @Nonnull @Override public RealApolloCall<T> httpCachePolicy(@Nonnull HttpCachePolicy.Policy httpCachePolicy) {
-    if (executed.get()) throw new IllegalStateException("Already Executed");
+    if (state.get() != IDLE) throw new IllegalStateException("Already Executed");
     return toBuilder()
         .httpCachePolicy(checkNotNull(httpCachePolicy, "httpCachePolicy == null"))
         .build();
   }
 
   @Nonnull @Override public RealApolloCall<T> responseFetcher(@Nonnull ResponseFetcher fetcher) {
-    if (executed.get()) throw new IllegalStateException("Already Executed");
+    if (state.get() != IDLE) throw new IllegalStateException("Already Executed");
     return toBuilder()
         .responseFetcher(checkNotNull(fetcher, "responseFetcher == null"))
         .build();
   }
 
   @Nonnull @Override public RealApolloCall<T> cacheHeaders(@Nonnull CacheHeaders cacheHeaders) {
-    if (executed.get()) throw new IllegalStateException("Already Executed");
+    if (state.get() != IDLE) throw new IllegalStateException("Already Executed");
     return toBuilder()
         .cacheHeaders(checkNotNull(cacheHeaders, "cacheHeaders == null"))
         .build();
   }
 
-  @Override public void cancel() {
-    canceled = true;
-    interceptorChain.dispose();
-    if (queryReFetcher.isPresent()) {
-      queryReFetcher.get().cancel();
+  @Override public synchronized void cancel() {
+    switch (state.get()) {
```

Comment:
@sav007 But as our state transitions involve mutating state (contained in multiple variables) inside the `if` you have sketched above, couldn't another thread see the new state flag before state transition side effects have finished?

Maybe I'm just not fully grasping your solution -- I won't have time to work on this until this weekend, so feel free to pick this up if you would like in the meantime.